shimmingtoolbox/coils/siemens_basis.py

```diff
@@ -0,0 +1,154 @@
+
+        Gist: given the stated nominal values, we can pick several arbitrary reference positions around the
+        origin/isocenter at which we know what the field *should* be, and use that to calculate the appropriate scaling
+        factor.
+
+        Returns:
+             numpy.ndarray:  1D vector of ``scaling_factors``
+
+        NOTE: The method has been worked out empirically and has only been tested for 2 Siemens Prisma systems.
+        E.g.re: Y, Z, ZX, and XY terms, their polarity had to be flipped relative to the form given by
+        ``shimmingtoolbox.coils.spherical_harmonics``.
+        """
+
+        [x_iso, y_iso, z_iso] = np.meshgrid(np.array(range(-1, 2)), np.array(range(-1, 2)), np.array(range(-1, 2)),
+                                            indexing='xy')
+        orders = np.array(range(1, 3))
+        sh = spherical_harmonics(orders, x_iso, y_iso, z_iso)
+        sh = reorder_to_siemens(sh)
+
+        n_channels = sh.shape[3]
+        scaling_factors = np.zeros(n_channels)
+
+        # indices of reference positions for normalization:
+        i_x1 = np.nonzero((x_iso == 1) & (y_iso == 0) & (z_iso == 0))
+        i_y1 = np.nonzero((x_iso == 0) & (y_iso == 1) & (z_iso == 0))
+        i_z1 = np.nonzero((x_iso == 0) & (y_iso == 0) & (z_iso == 1))
+
+        i_x1z1 = np.nonzero((x_iso == 1) & (y_iso == 0) & (z_iso == 1))
+        i_y1z1 = np.nonzero((x_iso == 0) & (y_iso == 1) & (z_iso == 1))
+        i_x1y1 = np.nonzero((x_iso == 1) & (y_iso == 1) & (z_iso == 0))
+
+        # order the reference indices like the sh field terms
+        i_ref = [i_x1, i_y1, i_z1, i_z1, i_x1z1, i_y1z1, i_x1, i_x1y1]
+
+        # distance from iso/origin to adopted reference point[units: mm]
+        r = [1, 1, 1, 1, np.sqrt(2), np.sqrt(2), 1, np.sqrt(2)]
+
+        # invert polarity
+        # Y, Z, ZX, and XY terms only(determined empirically)
+        sh[:, :, :, [1, 2, 4, 7]] = -sh[:, :, :, [1, 2, 4, 7]]
+
+        # scaling:
+        orders = [1, 1, 1, 2, 2, 2, 2, 2]
+
+        for i_ch in range(0, n_channels):
+            field = sh[:, :, :, i_ch]
+            gyro_ratio = 42.576
+            scaling_factors[i_ch] = gyro_ratio * ((r[i_ch] * 0.001) ** orders[i_ch]) / field[i_ref[i_ch]]
+
+        return scaling_factors
+
+    # Check inputs
+    if not (x.ndim == y.ndim == z.ndim == 3):
+        raise RuntimeError("Input arrays X, Y, and Z must be 3d")
+
+    if not (x.shape == y.shape == z.shape):
+        raise RuntimeError("Input arrays X, Y, and Z must be identically sized")
+
+    # Create spherical harmonics from first to second order
+    orders = np.array(range(1, 3))
+    spher_harm = spherical_harmonics(orders, x, y, z)
+
+    # Reorder according to siemens convention: X, Y, Z, Z2, ZX, ZY, X2-Y2, XY
+    reordered_spher = reorder_to_siemens(spher_harm)
+
+    # scale according to
+    # - 1 micro-T/m for *X,Y,Z* gradients (= 0.042576 Hz/mm)
+    # - 1 micro-T/m^2 for 2nd order terms (= 0.000042576 Hz/mm^2)
```

Comment:
I need to look into the physical situation behind this, but is this indicating that different channels have different units, or is this scaling to account for that difference? What are the final units for the returned coil profiles -- Hz?

Reply:
I found this handy [pdf](https://cbbi.udel.edu/wp-content/uploads/2017/01/Sys-DataManagement.pdf) which talks about the siemens console, on page 127, it shows a picture that shows how to perform shims on the console using the built in coils. I see it as an `Ax=-b` problem where A is the matrix of coil profiles with the siemens coefficients in micro-T/m for the first order and in micro-T/m^2 for the second order. `b` is the fieldmap to solve for in hz and `x` are the coefficients we are looking to input in the system. I'm also new to the subject so confirmation from @jcohenadad or @rtopfer would be great.

Reply:
@lcbMGH @po09i 
By design, the three 1st-order shims are linear fields (the X, Y and Z gradients).
E.g. Assume an initial field distribution that is on resonance everywhere (perfectly shimmed), were you to enter a value of "1" in the "X" row of the "Temporary" column of the Siemens 3D Shim adjustments page and press "Apply", the local field 1mm to the left of isocenter will be 1 nano-T greater than the field at isocenter (the origin where no field shift is introduced); another 1mm to the left, the field will be 2 nano-T. Equivalently, from the Larmor equation, dividing by gamma puts the shift in terms of Hz rather than Tesla.
Same concept applies to the 2nd-order terms, only the given units differ since the rate by which the field-shift increases/decreases moving away from isocenter is now quadratic rather than linear.
For any inductive shim coil, you're applying a current and introducing a certain amount of field shift. For the local "multi-coil" arrays, we're generally not designing them to meet a highly specific profile in the way the scanner shims are designed; i.e. spatial derivatives of the local coil profiles aren't constants: if it just so happened to be 1 micro-T/m somewhere, the figure will be entirely different somewhere. So, instead, we typically map out the field profiles and normalize them at 1 ampere (since the current circulating in a coil is what we actually control in an experiment). Siemens is likewise changing the current circulating in their shims, they've just opted to reframe the settings in this way since, in principal, it obviates the need to save the field profiles, which are supposed to conform to theoretical spherical harmonics, with the displayed units indicating the amplitudes/scaling values. (In practice, as one moves further and further away from isocenter, the actual Siemens profiles will deviate more and more from spec and their spatial derivatives won't conform reliably to the given constants).

The 3D Shim card also features the "Measure" & "Calculate" buttons, where "Measure" is acquiring _some_ form of field measurement and "Calculate" is performing _some_ type of optimization. There are different options available for the type of measurement, but the overall shim routine here is something of a black box. Surely it's along the lines of solving the least-squares `Ax=-b` problem, but the general consensus seems to be that manually acquiring a field map and performing the optimization offline does a better job than Siemens, so I'm not sure what they're doing exactly (and they weren't willing to reveal anything else about it when we asked a few years ago).